File: src/main/java/org/simbrain/network/subnetworks/FeedForward.java
/*
 * Part of Simbrain--a java-based neural network kit Copyright (C) 2005,2007 The
 * Authors. See http://www.simbrain.net/credits This program is free software;
 * you can redistribute it and/or modify it under the terms of the GNU General
 * Public License as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version. This program is
 * distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU General Public License for more details. You
 * should have received a copy of the GNU General Public License along with this
 * program; if not, write to the Free Software Foundation, Inc., 59 Temple Place
 * - Suite 330, Boston, MA 02111-1307, USA.
 */
package org.simbrain.network.subnetworks;

import org.simbrain.network.core.Network;
import org.simbrain.network.core.NeuronUpdateRule;
import org.simbrain.network.groups.Subnetwork;
import org.simbrain.network.matrix.NeuronArray;
import org.simbrain.network.matrix.WeightMatrix;
import org.simbrain.network.neuron_update_rules.LinearRule;
import org.simbrain.network.util.Direction;

import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.List;

import static org.simbrain.network.util.NetworkLayoutManagerKt.offsetNeuronGroup;

/**
 * A standard feed-forward network, where a succession of neuron groups and
 * synapse groups are organized into layers.
 *
 * @author Jeff Yoshimi
 */
public class FeedForward extends Subnetwork {

    /**
     * Space to put between layers.
     */
    private int betweenLayerInterval = 250;

    /**
     * Ordered reference to underlying array list.
     */
    private List<NeuronArray> layerList = new ArrayList<>();

    /**
     * Reference to input layer.
     */
    private NeuronArray inputLayer;

    /**
     * Reference to output layer.
     */
    private NeuronArray outputLayer;

    /**
     * Construct a feed-forward network.
     *
     * @param network             the parent network to which the layered network is being added
     * @param nodesPerLayer       an array of integers which determines the number of layers and
     *                            neurons in each layer. Integers 1...n in the array correspond
     *                            to the number of nodes in layers 1...n.
     * @param initialPosition     bottom corner where network will be placed.
     * @param inputNeuronTemplate the type of Neuron to use for the input layer
     */
    public FeedForward(final Network network, int[] nodesPerLayer, Point2D initialPosition,
                       final NeuronUpdateRule inputNeuronTemplate) {
        super(network);
        buildNetwork(network, nodesPerLayer, initialPosition, inputNeuronTemplate);
        super.events.onDeleted(nm -> {
            if (nm instanceof NeuronArray) {
                layerList.remove((NeuronArray) nm);
            }
        });
    }

    /**
     * Add the layered network to the specified network, with a specified number
     * of layers and nodes in each layer.
     *
     * @param network         the parent network to which the layered network is being added
     * @param nodesPerLayer   an array of integers which determines the number of layers and
     *                        neurons in each layer. Integers 1...n in the array correspond
     *                        to the number of nodes in layers 1...n.
     * @param initialPosition upper left corner where network will be placed.
     */
    public FeedForward(final Network network, int[] nodesPerLayer, Point2D initialPosition) {
        super(network);
        LinearRule rule = new LinearRule();
        rule.setLowerBound(0);
        buildNetwork(network, nodesPerLayer, initialPosition, rule);
    }

    /**
     * Create the network using the parameters.
     *
     * @param network             the parent network to which the layered network is being added
     * @param nodesPerLayer       an array of integers which determines the number of layers and
     *                            neurons in each layer. Integers 1...n in the array correspond
     *                            to the number of nodes in layers 1...n.
     * @param initialPosition     bottom corner where network will be placed.
     * @param inputNeuronTemplate the type of Neuron to use for the input layer
     */
    private void buildNetwork(final Network network, int[] nodesPerLayer, Point2D initialPosition,
                              final NeuronUpdateRule inputNeuronTemplate) {

        setLabel("Layered Network");

        inputLayer = new NeuronArray(network, nodesPerLayer[0]);
        addModel(inputLayer);
        layerList.add(inputLayer);

        // Memory of last layer created
        NeuronArray lastLayer = inputLayer;

        // Make hidden layers and output layer
        for (int i = 1; i < nodesPerLayer.length; i++) {
            NeuronArray hiddenLayer = new NeuronArray(network, nodesPerLayer[i]);
            addModel(hiddenLayer);
            layerList.add(hiddenLayer);
            offsetNeuronGroup(lastLayer, hiddenLayer, Direction.NORTH, betweenLayerInterval / 2, 100, 200);

            // Add weight matrix
            WeightMatrix wm = new WeightMatrix(getParentNetwork(), lastLayer, hiddenLayer);
            wm.randomize();
            addModel(wm);

            // Reset last layer
            lastLayer = hiddenLayer;
        }
        outputLayer = lastLayer;

    }

    public int getBetweenLayerInterval() {
        return betweenLayerInterval;
    }

    public void setBetweenLayerInterval(int betweenLayerInterval) {
        this.betweenLayerInterval = betweenLayerInterval;
    }

    @Override
    public String getName() {
        return null;
    }

    @Override
    public void onCommit() {
    }

    @Override
    public void update() {

        inputLayer.updateInputs();
        inputLayer.update();
        for (int i = 1; i < layerList.size()-1; i++) {
            layerList.get(i).updateInputs();
            layerList.get(i).update();
        }
        outputLayer.updateInputs();
        outputLayer.update();
    }

}

File: src/main/kotlin/org/simbrain/network/subnetworks/FeedForward.kt
package org.simbrain.network.subnetworks

import org.simbrain.network.core.Network
import org.simbrain.network.core.NeuronArray
import org.simbrain.network.core.WeightMatrix
import org.simbrain.network.core.randomizeBiases
import org.simbrain.network.util.Direction
import org.simbrain.network.util.offsetNetworkModel
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.util.stats.distributions.NormalDistribution
import java.awt.geom.Point2D

/**
 * A standard feed-forward network, as a succession of [NeuronArray] and [WeightMatrix] objects.
 *
 * @author Jeff Yoshimi
 */
open class FeedForward : Subnetwork {

    var betweenLayerInterval = 250

    /**
     * Ordered reference to [NeuronArray]'s maintained in [Subnetwork.modelList]
     */
    val layerList: MutableList<NeuronArray> = ArrayList()

    val wmList: MutableList<WeightMatrix> = ArrayList()

    lateinit var inputLayer: NeuronArray
        private set

    lateinit var outputLayer: NeuronArray
        private set

    constructor(): super()

    /**
     * @param parentNetwork Parent network
     * @param nodesPerLayer Integers 1...n correspond to number of nodes in layers 1..n
     * @param initialPosition Center location for network.
     */
    constructor(nodesPerLayer: IntArray, initialPosition: Point2D?): super() {
        inputLayer = NeuronArray(nodesPerLayer[0])
        inputLayer.label = "Input"
        addModel(inputLayer)
        layerList.add(inputLayer)

        // Memory of last layer created
        var lastLayer = inputLayer

        // Make hidden layers and output layer
        for (i in 1 until nodesPerLayer.size) {
            val hiddenLayer = NeuronArray(nodesPerLayer[i])
            if (i < nodesPerLayer.size - 1) {
                val numHiddenLayers = nodesPerLayer.size - 2
                hiddenLayer.label = if (numHiddenLayers == 1) "Hidden" else "Hidden $i"
            } else {
                hiddenLayer.label = "Output"
            }
            addModel(hiddenLayer)
            layerList.add(hiddenLayer)
            offsetNetworkModel(
                lastLayer,
                hiddenLayer,
                Direction.NORTH,
                (betweenLayerInterval / 2).toDouble(),
                100.0,
                200.0
            )

            // Add weight matrix
            val wm = WeightMatrix(lastLayer, hiddenLayer)
            wm.randomize()
            addModel(wm)
            wmList.add(wm)

            // Reset last layer
            lastLayer = hiddenLayer
        }
        if (initialPosition != null) {
            setLocation(initialPosition.x, initialPosition.y)
        }
        outputLayer = lastLayer
    }

    override val name: String
        get() = "Feedforward"

    override fun onCommit() {}

    /**
     * A supplied [randomizer] applies to weights and biases alike; the differently scaled defaults are used
     * only when none is given. Passing one used to have no effect at all, which quietly made even a seeded
     * distribution produce a different network every time.
     */
    override fun randomize(randomizer: ProbabilityDistribution?) {
        wmList.forEach { wm -> wm.randomize(randomizer ?: NormalDistribution(0.0, .1)) }
        (layerList - inputLayer).forEach {
            it.clear()
            it.randomizeBiases(randomizer ?: NormalDistribution(0.0, .01))
        }
    }

    context(Network)
    override fun accumulateInputs() {
        inputLayer.accumulateInputs()
    }

    context(Network)
    override fun update() {
        inputLayer.update()
        for (i in 1 until layerList.size - 1) {
            layerList[i].accumulateInputs()
            layerList[i].update()
        }
        outputLayer.accumulateInputs()
        outputLayer.update()
    }

    fun hiddenLayers() = layerList.drop(1).take(layerList.size-2)

    override fun toString(): String {
        val hiddenLayerSizes = hiddenLayers().map { it.size }
        val hiddenInfo = if (hiddenLayerSizes.isEmpty()) "None" 
                        else hiddenLayerSizes.joinToString(", ")
        return """
            Name: $displayName
            Type: Feed Forward Network
            Input Layer: ${inputLayer.size} neurons
            Hidden Layers: $hiddenInfo
            Output Layer: ${outputLayer.size} neurons
            Total Layers: ${layerList.size}
        """.trimIndent()
    }

    override fun copy(): FeedForward {
        val copy = FeedForward()
        copy.betweenLayerInterval = betweenLayerInterval

        // Copy input layer
        copy.inputLayer = inputLayer.copy()
        copy.inputLayer.label = "Input"
        copy.addModel(copy.inputLayer)
        copy.layerList.add(copy.inputLayer)

        // Memory of last layer created
        var lastLayer = copy.inputLayer

        // Copy hidden layers and output layer
        for (i in 1 until layerList.size) {
            val layer = layerList[i].copy()
            if (i < layerList.size - 1) {
                val numHiddenLayers = layerList.size - 2
                layer.label = if (numHiddenLayers == 1) "Hidden" else "Hidden $i"
            } else {
                layer.label = "Output"
            }
            copy.addModel(layer)
            copy.layerList.add(layer)

            // Maintain layout
            offsetNetworkModel(
                lastLayer,
                layer,
                Direction.NORTH,
                (betweenLayerInterval / 2).toDouble(),
                100.0,
                200.0
            )

            // Copy weight matrix
            val wm = WeightMatrix(lastLayer, layer)
            wm.setMatrixValues(wmList[i-1].weights.clone())
            copy.addModel(wm)
            copy.wmList.add(wm)

            lastLayer = layer
        }

        copy.outputLayer = lastLayer
        return copy
    }
}
